compiler/rustc_mir/src/transform/promote_consts.rs

```diff
@@ -552,18 +552,25 @@ impl<'tcx> Validator<'_, 'tcx> {
 
                 match op {
                     BinOp::Div | BinOp::Rem => {
-                        if lhs_ty.is_integral() {
-                            // Integer division: the RHS must be a non-zero const.
-                            let const_val = match rhs {
-                                Operand::Constant(c) => {
-                                    c.literal.try_eval_bits(self.tcx, self.param_env, lhs_ty)
-                                }
-                                _ => None,
-                            };
-                            match const_val {
-                                Some(x) if x != 0 => {}        // okay
-                                _ => return Err(Unpromotable), // value not known or 0 -- not okay
+                        // The RHS must be a non-zero const.
+                        let const_val = match rhs {
+                            Operand::Constant(c) => {
+                                c.literal.try_eval_bits(self.tcx, self.param_env, lhs_ty)
+                            }
+                            _ => None,
+                        };
+                        let might_be_zero = match (const_val, lhs_ty.kind()) {
+                            (None, _) => true,
+                            (Some(x), ty::Int(_) | ty::Uint(_)) => x == 0,
+                            // Account for signed zero: mask out sign bit
+                            (Some(x), ty::Float(f)) => {
+                                let mask = !(1 << (f.bit_width() - 1));
```

Comment:
I don't quite follow what's going on here. Is `f` an f32 value? If so, you can just use `==0.0` and it'll pass if `f` is either `0.0` or `-0.0`
```rust
// this passes
assert!(0.0 == 0.0);
assert!(0.0 == -0.0);
```

Reply:
`f` is https://doc.rust-lang.org/nightly/nightly-rustc/rustc_type_ir/enum.FloatTy.html#

`x` is the bits of the float (plus some padding zeros on the left)